File: dotnet/DesignPatterns/StructuralDesignPatterns/ProxyDesignPattern/ProxyDesignPattern.Example1/RealVideoService.cs
namespace ProxyDesignPattern.Example1;

/// <summary>
/// RealSubject - The actual video streaming service.
/// Contains the real implementation that accesses the video database.
/// </summary>
public class RealVideoService : IVideoService
{
    private readonly Dictionary<string, Video> _videoDatabase;

    public RealVideoService()
    {
        // Simulated video database
        _videoDatabase = new Dictionary<string, Video>
        {
            { "V001", new Video("V001", "Introduction to Design Patterns", false) },
            { "V002", new Video("V002", "Advanced C# Techniques", true) },
            { "V003", new Video("V003", "SOLID Principles Explained", false) },
            { "V004", new Video("V004", "Microservices Architecture", true) }
        };
    }

    public Video? GetVideoInfo(string videoId)
    {
        Console.WriteLine($"[RealService] Fetching video info for: {videoId}");
        
        if (_videoDatabase.TryGetValue(videoId, out var video))
        {
            return video;
        }
        
        return null;
    }

    public void StreamVideo(string videoId)
    {
        Console.WriteLine($"[RealService] Loading video content from server...");
        Thread.Sleep(1000); // Simulate network latency

        if (_videoDatabase.TryGetValue(videoId, out var video))
        {
            video.Content = $"[Binary video data for '{video.Title}']";
            Console.WriteLine($"[RealService] Streaming: {video.Title}");
            Console.WriteLine($"[RealService] Content loaded successfully");
        }
    }
}

File: nodejs/DesignPatterns/StructuralDesignPatterns/ProxyDesignPattern/ProxyDesignPattern.Example1/src/RealVideoService.ts
import { IVideoService } from './IVideoService';
import { Video } from './Models';

export class RealVideoService implements IVideoService {
  private readonly videoDatabase = new Map<string, Video>([
    ['V001', new Video('V001', 'Introduction to Design Patterns', false)],
    ['V002', new Video('V002', 'Advanced TypeScript Techniques', true)],
    ['V003', new Video('V003', 'SOLID Principles Explained', false)],
    ['V004', new Video('V004', 'Microservices Architecture', true)],
  ]);

  getVideoInfo(videoId: string): Video | null {
    console.log(`[RealService] Fetching video info for: ${videoId}`);
    return this.videoDatabase.get(videoId) ?? null;
  }

  streamVideo(videoId: string): void {
    console.log('[RealService] Loading video content from server...');
    const video = this.videoDatabase.get(videoId);
    if (video) {
      video.content = `[Binary video data for '${video.title}']`;
      console.log(`[RealService] Streaming: ${video.title}`);
      console.log('[RealService] Content loaded successfully');
    }
  }
}
